### ISSUE TITLE
Execute a TaskRun without having to create a Task

### ISSUE THREAD

**Issue by ImJasonH** [knative/build-pipeline#249]
Related to #147

## Expected Behavior

As a former Build user, I should be able to create a single resource that executes requested steps and can be watched for updates.

## Actual Behavior

If I want to run some steps, I first have to define a `Task` that specifies those steps, then invoke it with a `TaskRun` that references that `Task`.

## Additional Info

One possible solution would be to have the `TaskRun` type take either a `taskRef` or a `taskSpec` (but not both, or neither). When a `TaskRun` is requested with a `taskSpec`, it would perform the same steps as if a `Task` had been created with the given spec, then invoked by the `TaskRun`, without creating a `Task` as a side effect. Tasks defined in this way can't specify parameters, but can specify inputs. Resources might also need to be able to be defined without referencing a pre-existing CRD resource.

**Comment by pivotal-nader-ziada** [comment]
@ImJasonH @bobcatfish Its not clear what value `pipeline` gets from this feature. This makes us maintain `build` style of executing steps along with `Task`, `TaskRun` using the same code base which would lead to more complexity. Is it better to help users migrate their existing `builds` into `tasks` and ask them to run them using `TaskRun`? if we feel the concept of `TaskRun` is too much for one-off tasks, maybe we should consider the option of allowing users to run Tasks in general, but its the two different ways of running things that concerns me.

**Comment by pivotal-nader-ziada** [comment]
good point @ImJasonH, the more I think about it, the more I think Task + TaskRun for one off builds or simple building of an app might be a little too much for a lot of use cases.
